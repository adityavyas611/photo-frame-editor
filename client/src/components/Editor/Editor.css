.editor {
    height: 100%;
}

.editor-canvas {
    align-items: center;
    display: flex;
    height: 100%;
    justify-content: center;
}

.uploaded-image {
    margin-top: 20px;
    max-height: 65%;
    max-width: 65%;
}