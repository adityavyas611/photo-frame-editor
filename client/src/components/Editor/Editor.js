import React from 'react';
import Toolbar from '../Toolbar/Toolbar';
import './Editor.css';

const Editor = ({imgData: {name}, imgUrl}) =>  {
    return (
        <div className="editor">
            <div className="editor-canvas">
                <img alt={name} src={imgUrl} className="uploaded-image" />
            </div>
            <Toolbar/>
        </div>
    );
};

export default Editor;