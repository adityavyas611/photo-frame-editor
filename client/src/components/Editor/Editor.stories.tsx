import React from 'react';
import Editor from './Editor';

export default {
  component: Editor,
  title: 'Editor',
};

const imageData = {
  name: 'Aditya'
};

export const editor = () => <Editor imgData={imageData} imgUrl="https://www.online-image-editor.com/styles/2019/images/power_girl.png" />;

editor.story = {
  name: 'to Editor',
};
