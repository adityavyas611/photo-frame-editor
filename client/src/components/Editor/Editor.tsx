import React, { useState, useEffect, useRef, RefObject } from 'react';
import Cropper from 'cropperjs';
import Toolbar from '../Toolbar/Toolbar';
import './Editor.css';
import '../Toolbar/Toolbar.css';
import 'cropperjs/dist/cropper.css';

type Props = {
	imgData: any,
	imgUrl: string | ArrayBuffer | null,
};

const Editor = ({ imgData, imgUrl }: Props) => {
	const [cropping, setCropping] = useState(false);
	const [cropped, setCropped] = useState(false);
	const [cropper, setCropper] = useState<Cropper>();
	const [frame, setFrame] = useState(false);

	const imageRef = useRef<HTMLImageElement>(null);
	const frameRef = useRef<HTMLImageElement>(null);

	const intitalizeCropper = (ref: RefObject<HTMLImageElement>) => {
		if (ref && ref.current) {
			const currentCropper = new Cropper(ref.current, {
				autoCrop: false,
				dragMode: 'move',
				background: false,
				minContainerWidth: 1000,
				minContainerHeight: 600,
				toggleDragModeOnDblclick: false,
				crop: ({ detail }) => {
					if (detail.width > 0 && detail.height > 0) {
						setCropping(true);
					}
				},
			});
			setCropper(currentCropper);
		}
	};

	const stopCropper = () => {
		if (cropper) {
			cropper.destroy();
		}
	};

	const cropImage = () => {
		if (cropping && cropper && imageRef && imageRef.current) {
			imageRef.current.src = cropper.getCroppedCanvas({}).toDataURL();
			setCropped(true);
			setCropping(false);
		}
		stopCropper();
	};

	const clearCrop = () => {
		if (cropper) {
			cropper.clear();
			setCropping(false);
		}
	};

	const showFrame = () => {
		setFrame(!frame);
	};

	useEffect(() => {
		if (imgUrl && imageRef) {
			stopCropper();
			intitalizeCropper(imageRef);
		}
	}, [imgUrl]);

	useEffect(() => {
		if (frame && frameRef) {
			stopCropper();
			intitalizeCropper(frameRef);
		}
	}, [frame]);

	const handleToolbarClick = (event: any) => {
		const { target } = event;
		const buttonAction = target.getAttribute('data-action');
		switch (buttonAction) {
			case 'move':
			case 'crop':
				cropper && cropper.setDragMode(buttonAction);
				break;
			case 'rotate-left':
				cropper && cropper.rotate(-90);
				break;
			case 'rotate-right':
				cropper && cropper.rotate(90);
				break;
			case 'scale':
				cropper && cropper.scale(-cropper.getData().scaleX, -cropper.getData().scaleY);
				break;
			case 'crop-tick':
				cropImage();
				break;
			case 'crop-cancel':
				clearCrop();
				break;
			default:
		}
	}
	const { name } = imgData;
	return (
		<div className="editor">
			<div className="editor-canvas">
					<img alt={name} src={imgUrl ? String(imgUrl) : ''} className={frame ? "frame-image" : "uploaded-image"} ref={imageRef} />
					{frame ? <img alt="frame" src="./images/goldframe.png" ref={frameRef} className="image-frame" /> : null}
			</div>
			{!cropped ? <Toolbar handleDataAction={handleToolbarClick} isCropping={cropping} />
				: <div className="toolbar">
					<img className="toolbar-button" src="./images/frame.png" alt="Select Frame" title="Select Frame" onClick={showFrame} />
				</div>}
		</div>
	);
};

export default Editor;
