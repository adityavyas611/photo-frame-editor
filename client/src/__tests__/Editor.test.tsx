import React from 'react';
import {shallow, configure} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import Editor from '../components/Editor/Editor';

configure({ adapter: new Adapter() })

describe('Create Editor', () => {
  it('should render editor component properly', () => {
    const wrapper = shallow(<Editor imgData={{}} imgUrl="" />)
    expect(wrapper).toMatchSnapshot();
  });
});