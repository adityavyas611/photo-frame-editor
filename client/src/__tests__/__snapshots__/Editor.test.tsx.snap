// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Editor should render properly 1`] = `ShallowWrapper {}`;
